### PR TITLE
Allow to name existentials in pattern-matching

### PR DESCRIPTION
**Pull request by garrigue**
This attempts to implement a solution suggested by @yallop in #7074 to the problem of naming existentials in pattern-matching.
For the sake of simplicity(?) the syntax is:
```ocaml
type _ ty = Int : int ty
type dyn = Dyn : 'a ty * 'a -> dyn
let f = function Dyn (type a) (w, x : a ty * a) -> ignore (x : a)
```
I.e. the type is given as a tuple rather than on individual arguments.
Ellipses are allowed, but all existentials must be matched.

### CODE REVIEW

**Comment by gasche** [comment]
@trefis pinged me about this PR just now. I like the proposed syntax, *much* better than the one in #9579. There is broad consensus in #7074 that this syntax is acceptable (@yallop, @lpw25 and myself supported this syntax, @CraigFe had solid arguments against the other proposal, etc.).

I remain of the opinion that it should ideally be possible to have something "more local" in the annotating-the-argument parts, for example `C (type a) ((x : a), f)` should be accepted just as well as `C (type a b) (x, f : a * (a -> b))`, but this could come later as a relaxation of the current feature. Even in that relaxed world, the syntax you just proposed would be accepted and make sense.

(I'm only semi-happy with the `*` which entertains the confusion between tuple types and multi-argument constructors, but it is consistent with the GADT declaration syntax so probably fine.)

**Comment by gasche** [comment]
I think it would be nice to move forward: we have a syntax we like, now we need to check that the implementation is reasonably consensual (maybe @trefis or @lpw25 could chime in with their perspective here), and seriously consider this for inclusion.

File: parsing/ast_iterator.ml
Comment (lpw25) [inline parsing/ast_iterator.ml]: Should probably iterate on the `string loc`s too

File: parsing/ast_mapper.ml
Comment (lpw25) [inline parsing/ast_mapper.ml]: Should probably map on the `string loc`s too

File: parsing/printast.ml
Comment (lpw25) [inline parsing/printast.ml]: Better not to just ignore them in the printer.

File: typing/ctype.ml
Comment (lpw25) [inline typing/ctype.ml]: I see no obvious benefit to making these parameters optional. It just makes it likely someone will forget one. Making them labelled is good though.

File: typing/typecore.ml
Comment (lpw25) [inline typing/typecore.ml:1631]: It's not immediately clear to me why this call is to `Env.enter_type` but the one inside of `Ctype.instance_constructor` is to `Env.add_local_type`. I think it would be good if we could find a way to factor these bits of code together.

File: typing/untypeast.ml
Comment (lpw25) [inline typing/untypeast.ml]: I think that we should add something to the `Typedtree` to hold the names of the existientials and the typed type expression from the constraint. Then we should use that here rather than just assuming it is `None`.

**Review by lpw25 — APPROVED** [review]
LGTM. A few small things to improve, but the code looks correct to me. Well done everyone involved for finding something we can all agree to.

File: parsing/parser.mly
Comment (gasche) [inline parsing/parser.mly]: Now that we have the Menhir parser, we can use names instead of positions. We didn't convert existing rules to make the conversion less invasive, but I would be in favor of using them when appropriate for new rules. Here I find `$4`, `$9` and `$7` difficult to read, could we use names instead?

```ocaml
| constr=mkrhs(constr_longident)
  LPAREN TYPE newtypes=lident_list RPAREN
  LPAREN pat=pattern COLON ty=core_type RPAREN
  { Ppat_construct(constr, Some(newtypes, core_type), Some pattern) }
```

File: parsing/parsetree.mli
Comment (gasche) [inline parsing/parsetree.mli]: The comment should be updated here, to reflect the new constructor form.

Comment (gasche) [reply parsing/parsetree.mli]: Are we sure that adding a new argument is the right choice, though? If I understand correctly, the new syntax only allows annotating constructor parameters, not the return type (the point is to name existential variables which do not occur in the return type anyway), so we can never have Some annotation if we do not also have Some parameter. To me this suggests that we could have

```ocaml
| Ppat_constructor of Longident.t loc * (pattern * quantified_type option) option
```

which has the benefit of resulting in a sensibly less invasive patch, as the arity of the constructor does not change.

File: parsing/ast_iterator.ml
Comment (garrigue) [reply parsing/ast_iterator.ml]: Done

File: parsing/printast.ml
Comment (garrigue) [reply parsing/printast.ml]: Done

File: typing/ctype.ml
Comment (garrigue) [reply typing/ctype.ml]: This is basically a smart constructor, and I expect smart constructors to be smart, i.e. provide sensible defaults.
This function is only used in a small number of occurrences, and I think having defaults here avoids cluttering.

File: typing/typecore.ml
Comment (garrigue) [reply typing/typecore.ml:1631]: Good point. I have switched all creations of local types to `enter_type`, using only `add_local_type` for redefinitions.

File: typing/untypeast.ml
Comment (garrigue) [reply typing/untypeast.ml]: Done

File: parsing/parser.mly
Comment (garrigue) [reply parsing/parser.mly]: Done

File: parsing/parsetree.mli
Comment (garrigue) [reply parsing/parsetree.mli]: Done

**Comment by garrigue** [comment]
After many weeks and two rebases, I think this PR is now ready for inclusion.
I also added an entry for the manual.
@lpw25 already gave his approval, but since there are a number of changes, it would be good if at least one of you could give a fresh look.

File: typing/ctype.ml
Comment (gasche) [reply typing/ctype.ml]: But `Location.none` is not a *smart* or *sensible* default for an optional location argument, it is the guarantee that some user down the line will get a worse experience because the location is not shown.

File: typing/typedtree.mli
Comment (gasche) [inline typing/typedtree.mli]: The comment here should be updated, just like in parsetree.mli (thanks for that change, by the way).

File: typing/ctype.ml
Comment (lpw25) [reply typing/ctype.ml]: I don't want to hold the patch up over a triviality, but I pretty strongly disagree with this. Using `Btype.lowest_level` for the expansion scope would be a subtle soundness bug at some of the uses of this function, and using `Location.none` at some other uses would cause us to fail to produce some usage warnings. It's much more important to avoid those bugs than to avoid cluttering the call sites of a rarely used function. I'm less worried about the `manifest` parameter having a default since that one should be much harder to forget.

File: typing/tast_iterator.ml
Comment (lpw25) [inline typing/tast_iterator.ml]: I think this should be iterating over the new argument.

File: typing/typecore.ml
Comment (lpw25) [inline typing/typecore.ml]: This line makes me nervous since we are not doing anything with `ty` in this case. Does the code behave correctly on:
```ocaml
| None (type a) (_ : a) -> ...
```

File: typing/printpat.ml
Comment (lpw25) [inline typing/printpat.ml:59]: Should maybe also be `None` in the new argument? Probably can't happen, but more obviously correct.

File: typing/printpat.ml
Comment (lpw25) [inline typing/printpat.ml]: Should be `None` in the new argument, this one can actually happen.

File: typing/printpat.ml
Comment (lpw25) [inline typing/printpat.ml]: Should be `None` in the new argument.

File: typing/printpat.ml
Comment (lpw25) [inline typing/printpat.ml]: Should be `None` in the new argument.

File: typing/printpat.ml
Comment (lpw25) [inline typing/printpat.ml]: Should be `None` in the new argument.

**Review by lpw25 — APPROVED** [review]
Spotted a couple of small things to change and maybe a small bug. Good to go once they are addressed.

File: typing/typecore.ml
Comment (garrigue) [reply typing/typecore.ml]: We are unifying `ty` with `ty_args` just below. This is the intended behavior.

File: typing/printpat.ml
Comment (garrigue) [reply typing/printpat.ml:59]: Not really more correct: it would result in printing `C (type a) ( : _)`.
Or do you mean that we should print something clearly wrong if the invariant is broken?

File: typing/printpat.ml
Comment (garrigue) [reply typing/printpat.ml]: Ok

File: typing/ctype.ml
Comment (garrigue) [reply typing/ctype.ml]: For `Location.none`, I beg to disagree. This is the only possible default. This function is called from places where no location is available, and it would make no sense to make that more explicit. We're not talking of a 3rd party API but of an internal function of the compiler.
For `expansion_scope`, you have a point, and looking back at the code it seems actually needlessly complicated.
I.e., the `expansion_scope` is used only when one expands an abbreviation, so it makes no sense to give it if there is no manifest. So I have changed the arguments to `?loc` and `?manifest_and_scope`, which make much more sense. The testsuite goes through, demonstrating that setting it explicitly was useless.

File: typing/printpat.ml
Comment (garrigue) [reply typing/printpat.ml]: Done

File: typing/printpat.ml
Comment (garrigue) [reply typing/printpat.ml]: Done

File: typing/printpat.ml
Comment (garrigue) [reply typing/printpat.ml]: Done

**Comment by gasche** [comment]
Note: I think that the feature is very useful/important and that the design we ended up with is good. Thanks a lot @garrigue  for this work.

File: parsing/parsetree.mli
Comment (alainfrisch) [inline parsing/parsetree.mli]: Sorry, I'm late in the discussion, but shouldn't we encode a regular Ppat_constraint to keep the core_type, i.e. just use:

````ocaml
  Ppat_construct of loc * (pattern * string loc list) option
````

(with an empty list when there is no `(type ...)`)  Requiring to have the inner pattern be a Ppat_constraint when the list is not empty is ok, imo.

I'm concerned that the current representation changes the way an existing constraint would be represented when one adds a type binding (`C (P : T)` --> `C (type a) (P : T')`).

Can the type-checked behave differently on `C (type a) (P : T)` and `C (type a) ((P : T) : T)`?

Btw, is it an error to bind type names which are not used in the pattern?

Comment (gasche) [reply parsing/parsetree.mli]: I thought about suggesting such nesting, but:
- Currently the annotation is interpreted in the context of the type variables in a very specific way, so it makes sense to keep the two pieces together at the constructor site.
- Even with the representation you propose, my understanding is that the type-checker would have to lookup the `Tpat_constraint` node and use its payload specifically (not go through the usual `Tpat_constraint` path in this case), so you would not get the same-behavior guarantees that you are asking about.

The handling of existential variables is very specific because, morally, their actual binding is computed by unifying the annotation with the constructor return type, and then making them rigid for the rest of the type-checking. This is not the behavior you would get if you checked the whole subpattern before rigidifying (I think that behavior would make sense and we discussed it earlier, but @garrigue says it is harder to implement), nor of course the behavior you would get by rigidifying before checking the annotation or using flexible variables.

Comment (gasche) [reply parsing/parsetree.mli]: Note: the previous discussion I was referring to is starting (approximatively) at https://github.com/ocaml/ocaml/issues/7074#issuecomment-630847726 .

File: typing/typecore.ml
Comment (lpw25) [reply typing/typecore.ml]: But not in the case where `ty_args` is the empty list. Which means that the following code compiles without issue on this branch:
```ocaml
[@@@warning "-28"]

let () =
  match None with
  | None (type a) (_ : a * int) -> ()
  | Some _ -> ()
```

I suggest that we prevent this example by changing line 1572 to only allow `_` to mean multiple or zero arguments when there is no type constraint. Then this line can be replaced by `| [] -> assert false` and it will be clear that `ty` is actually used on all paths.

File: typing/printpat.ml
Comment (lpw25) [reply typing/printpat.ml:59]: Sorry, I was confusing this bit of code with the stuff below, and thought it was related to handling `::`.

**Comment by lpw25** [comment]
You replied "done" to a few remarks, but I don't see the changes. A forgotten push maybe?

File: typing/tast_iterator.ml
Comment (garrigue) [reply typing/tast_iterator.ml]: Done

**Comment by garrigue** [comment]
> You replied "done" to a few remarks, but I don't see the changes. A forgotten push maybe?

No, this should be there. They are marked as outdated, so the code should have changed.
If there is still a problem, please add a new comment.

File: typing/typecore.ml
Comment (garrigue) [reply typing/typecore.ml]: Ok, I had misunderstood. Done as suggested.

File: parsing/parsetree.mli
Comment (alainfrisch) [reply parsing/parsetree.mli]: Thanks @gasche for the pointer.

 - If the type-checking is not an extension of what happens for `C (P : T)`, reusing the same concrete syntax can really be misleading, no?

 - Independently of what the type checker does, having different representation for the T in `C (type a) (P : T)` and in `C (P : T)` is confusing for people writing syntactic tools.  I'm not shocked that, short of having a cleaner solution, the type-checker does something special with Ppat_constraint on top of the pattern when existential type names are bound.  We might be able to change the behavior of the type-checker later, while keeping the same syntax and Parsetree representation.

Comment (gasche) [reply parsing/parsetree.mli]: > If the type-checking is not an extension of what happens for C (P : T), reusing the same concrete syntax can really be misleading, no?

I think that the two readings of `C (P : T)` (as a constraint, and as a return-type-annotated-constructor pattern with no existential variables bound) do have the same meaning. What I meant is that if we think of the type-checking of a pattern `P` as generating an inference constraint (or some other "result", for example in an non-inferred world taking an scrutinee type and returning types for the bound variables), then the result of type-checking of `C (type a...) (P : T)` cannot be computed in a modular lay, as an action on the result of the type-checking of `(P : T)`.

> Independently of what the type checker does, having different representation for the T in C (type a) (P : T) and in C (P : T) is confusing for people writing syntactic tools. 

This is a good point.

Comment (garrigue) [reply parsing/parsetree.mli]: Answering @alainfrisch 's concerns:
* Except for the special handling of variables introduced by `(type a ...)`, the code is supposed to do exactly the same thing as `Tpat_constraint`.
* In particular, `C (type a) (P : T)` and `C (type a) ((P : T) : T)` should be strictly equivalent. I.e., `a` is bound by the external type annotation, and reusing in the internal one should give the same type.
* It is not an error to introduce a type name without binding it. It just has the same meaning as a usual `(type a)`.
* However, one must bind all existentials arising from the constructor, with distinct names, and only once.

Of course, the first part is only true if there is no bug in the code, which I hope is the case, but the complexity of `Tpat_construct` for GADT constructors makes the comparison non-trivial.
For the last point, here are some examples:
```ocaml
# type dyn2 = Dyn : 'a * 'b -> dyn2;;
# function Dyn (type a b) (_ : a * b) -> ();;
- : dyn2 -> unit = <fun>
# function Dyn (type a) (_ : a * _) -> ();;
Error: This type does not bind all existentials in the constructor:
         type a. a * 'a
# function Dyn (type a) (_ : a * a) -> ();;
Error: This type does not bind all existentials in the constructor:
         type a. a * a
# type dynp = Dynp : 'a * 'a -> dynp;;
type dynp = Dynp : 'a * 'a -> dynp
# function Dynp (type a b) (_ : a * b) -> ();;
Error: This pattern matches values of type a * b
       but a pattern was expected which matches values of type a * a
       Type b is not compatible with type a 
```

Comment (alainfrisch) [reply parsing/parsetree.mli]: >  then the result of type-checking of C (type a...) (P : T) cannot be computed in a modular lay, as an action on the result of the type-checking of (P : T).

But that's not something unexpected for a binding construction, right?

If I extrapolate @garrigue's answer, the code currently used for `C (type ...) (P : T)` should do the same thing when applied to an empty list of binders as the existing `C (P : T)` (modulo bugs).  It's even more tempting to use the representation I proposed.

Note that expecting a specific root structure in the Parsetree is not something unheard of : n-ary constructors are parsed and represented in the Parsetree as tuple, and the type-checker reacts in a special way to those Ppat_tuple/Pexp_tuple under n-ary constructor (this is even less local than what I propose, which depends on the presence of a non-empty of binders, i.e. a local syntactic property).

> This type does not bind all existentials in the constructor:

Why is this rejected?  Could those remaining existentials be treated as they are today, when there is no binding?

Comment (lpw25) [reply parsing/parsetree.mli]: Personally, I prefer the current representation. The proposed change adds yet another hidden invariant to the parsetree, making it harder to reliably generate valid parsetrees in ppxes. It's similar to the current encoding of `let x : 'a. 'a -> 'a = ...` which has been a persistent cause of pain to ppx authors.

Comment (gasche) [reply parsing/parsetree.mli]: I think that Alain's proposal would make a lot of sense if we had a modular desugaring of `C (type a ..) (P : T)` into `(type a ...) P` (which introduces open existential types with can only be unified with distinct pure existentials within P and are opaque to the outside) and `C P` and `(P : T)`. But we are not currently able to implement this richer (and I think nicer) semantics today.

If we had good reasons to believe that we would be able to have such open-existential-types in the medium term, then it would make sense to argue for this representation now (even if it is somewhat of a lie today). But so far I was the only one to advocate for this extended feature, and I'm not expert enough to be confident that it even makes sense.

Comment (alainfrisch) [reply parsing/parsetree.mli]: >  The proposed change adds yet another hidden invariant to the parsetree, making it harder to reliably generate valid parsetrees in ppxes.

It's a matter of perspective.  I tend to think that `C (type a ...) P` is a type error when P is not a constraint, not a syntax error, so it's really not a matter of generating valid or invalid parstrees in ppx, or having hidden invariants (unless you consider all type-checking errors as hidden invariants!).  Moreover, it would be possible to later support putting constraints deeper in the pattern, and this would just be an improvement of the type-checker, with no need to change the syntax (e.g. it should be straightforward to accept `C (type a) ((P1 : t1), (P2 : t2))`, understood as `C (type a) ((P1, P2) : t1 * t2)`).

Comment (lpw25) [reply parsing/parsetree.mli]: If you're going to take that perspective then I really think the parser needs to accept things like `C (type a) 4` and we should emit the error from the type-checker instead.

Comment (gasche) [reply parsing/parsetree.mli]: Is there a possibility that we could postpone this discussion for a later PR, and merge this one with this question delayed? If I was in @garrigue's stead I would start getting slightly irritated by the back-and-forth.

Comment (alainfrisch) [reply parsing/parsetree.mli]: > If you're going to take that perspective then I really think the parser needs to accept things like C (type a) 4 and we should emit the error from the type-checker instead.

Yes, that's what I'm proposing.  I think it might give better error messages, btw.

> Is there a possibility that we could postpone this discussion for a later PR, and merge this one with this question delayed? If I was in @garrigue's stead I would start getting slightly irritated by the back-and-forth.

I did not follow the latest developments in the ppx ecosystem, but last time I checked, there was relatively high cost associated to changing the Parsetree definition.  If we can avoid having to adapt tooling for an intermediate version, that might be better.

But I take your point.  I came late to the discussion, so I'm not in a good position to block it if everyone else is ok with the current state.

Comment (gasche) [reply parsing/parsetree.mli]: I think that your argument does have merit (@lpw25 seems against and @garrigue has not spoken yet, but this is something I have considered as well and would find somewhat natural). We should take the time to discuss it, but maybe in a non-blocking way indeed.

Comment (gasche) [reply parsing/parsetree.mli]: (When I say "merge this one", I'm not actually sure that all pending questions have been resolved, this is for @lpw25 to confirm. So it may not be the unique blocker *yet*.)

Comment (lpw25) [reply parsing/parsetree.mli]: I'm fine with either the current version or the version where the check is completely moved to the type-checker and the parser is happy to parse patterns with existential binders and no type constraints. I dislike any version where what the type can represent and what the parser can produce are not the same.

File: typing/ctype.ml
Comment (lpw25) [reply typing/ctype.ml]: I disagree about `loc`, but that one worries me much less than `expansion_scope`, so I'm happy for it to be merged in its current state.

**Review by lpw25 — APPROVED** [review]

File: parsing/parsetree.mli
Comment (garrigue) [reply parsing/parsetree.mli]: I just realize that it is wrong to say that
```
C (type a) (pats : ty)
```
is equivalent to
```
C (pats : ty)
```
when `a` occurs nowhere: if `C` has arity 2 or more, `pats` must syntactically be a list of patterns, and the latter would not be typable. So, sorry, this is really a new syntax.
@gasche indeed suggested that it would be nicer to allow type constraints on each argument, but I answered that it would be difficult to implement it cleanly.
I will still give it a try, just to see how it would work.

Comment (alainfrisch) [reply parsing/parsetree.mli]: So, IIUC, for a pattern with 2 arguments:
  - `C _` is allowed but `C (type a) _` is rejected(?).
  - `C (type a) ((x, y) : t)` can be allowed (for a correct type) (?) but `C ((x, y) : t)` is always rejected.

Is that right? Are there other differences?

The situation with n-ary arguments was already quite confusing, I fear this is getting even worse...

**Comment by garrigue** [comment]
I have create another PR implementing(?) @alainfrisch 's idea (#10180).
Tell me which you prefer.
I don't like the syntax for type annotations, but the problem is indeed with ocaml syntax.

**Comment by garrigue** [comment]
I have merged most of the suggestions here and in #10180 and rebased.
* the parsetree represention uses a `Ppat_constraint`, which is checked in `Typecore`
* type annotations are only allowed on the argument seen as a tuple, but this is now also allowed for normal constructors too

**Comment by garrigue** [comment]
I'm getting a compilation error in the flambda backend, but I don't see how this can be related to this PR.
Will probably have to wait until this is fixed and rebase.

**Comment by garrigue** [comment]
Note also that neither of these approaches work for GADT constructors with an inline record.
The problem is not just that we are not allowed to name the record, but that there is no way to determine the order of the existentials it introduces.
A solution would be to allow typing a part of the arguments, delaying the GADT constructors they may contain, then check that the existentials are bound, and only then restart typing those eventual GADT constructors.
But this can probably wait for a future PR...

**Comment by garrigue** [comment]
I'm really confused about what is happening. Apparently the bug comes from this PR, but the error message is:
```
./boot/ocamlrun ./boot/ocamlc -g -nostdlib -I boot -use-prims runtime/primitives -strict-sequence -principal -absname -w +a-4-9-40-41-42-44-45-48-66 -warn-error A -bin-annot -safe-string -strict-formats -I utils -I parsing -I typing -I bytecomp -I file_formats -I lambda -I middle_end -I middle_end/closure -I middle_end/flambda -I middle_end/flambda/base_types -I asmcomp -I asmcomp/debug -I driver -I toplevel -c middle_end/flambda/simple_value_approx.ml -I middle_end/flambda
Fatal error: exception Invalid_argument("output_value: functional value")
make[2]: *** [middle_end/flambda/simple_value_approx.cmo] Error 2
```
Very confusing.
I hope it would disappear with a bootstrap, but actually this is even worse: before it occurred only when compiling with ocamlopt, but after bootstrap with ocamlc too.

**Comment by garrigue** [comment]
I could eventually fix the problem by being more conservative, but there is a mysterious interaction at work here.

**Comment by garrigue** [comment]
OK, I found the reason: I had forgotten to reinstate processing of the annotation `core_type` in `Tast_mapper` and `Tast_iterator`. It seems that this is used somewhere, and resulted probably in a malformed typedtree. I still do not understand how this can lead to outputting a functional value, but at least the original cause is now clear.

I'm still keeping the conservative approach for already valid annotations, as removing a `Ppat_constraint` node could change the behavior of `build_as_type` for instance.

**Comment by lpw25** [comment]
> I  still do not understand how this can lead to outputting a functional value, but at least the original cause is now clear.

The typed AST mapper is used for iterating over all the environments in the AST and calling [Env.keep_only_summary] on them before outputing the cmt file.

**Comment by garrigue** [comment]
Thanks for the explanation, this now makes sense.

Is everybody happy with this last iteration of the PR?
If nobody protests, I'm going to merge it tomorrow, as syntax PRs are a pain to rebase...

File: parsing/parsetree.mli
Comment (garrigue) [reply parsing/parsetree.mli]: Not quite correct.
* `C _` is allowed, but `C (type a) _` is rejected by the type checker (not the parser) independently of the number of arguments. The `(type a)` notation inside patterns is restricted to existential variables, and requires to bind all of them, which you cannot do without a type annotation, hence the rejection.
* Annotations are always allowed for multiary types (the internal parenthesis are optional).
```ocaml
type ('a, 'b) pair = Pair of 'a * 'b;;
type t = int * bool;;
let f = function Pair ((x,y) : t) -> x;;
val f : (int, bool) pair -> int = <fun>
type closure = Clos : ('a -> int) * 'a -> closure;;
type 'a app = ('a -> int) * 'a;;
let f = function Clos (type a) ((x,y) : a app) -> x y;;
val f : closure -> int = <fun>
```
The result looks more regular to me.

**Comment by Octachron** [comment]
This is minor, but am I right to think that there is no support for naming existentially quantified row variable:
```ocaml
type t = X: [> `A | `B ] -> t
let f (X (type x) (v:??)) = ()
```
?

**Comment by garrigue** [comment]
> This is minor, but am right to think that there is no support for naming existentially quantified row variable:

Very true, but this is an orthogonal problem: there is no support to quantify row variables using the `type a. <m : int; ..> as a` notation either. This is future work, probably after implementing full support for refining existentials row variables (not available yet).

**Comment by garrigue** [comment]
Thanks to everybody in the discussion, and good annotating!

**Comment by johnwhitington** [comment]
The directory `manual/manual` was moved to `manual/src` by https://github.com/ocaml/ocaml/commit/2aeb55a06b87203a247bfacbff981383b8dcab0e a few days before your merge, but your merge has recreated `manual/refman/exten_camltex.tex` in `manual/manual`, orphaning it. The full manual build now fails on trunk (although `make html` seems to work, luckily).

cc @octachron

**Comment by Octachron** [comment]
Thanks for the notice! This should be fixed in trunk.

@garrigue ,  to be safe, on the manual side, was there only a new paragraph in the  GADT section?

**Comment by garrigue** [comment]
Yes, that’s all I added.
I thought I understood the change of the manual, but this PR went through too many rebases.

**Comment by Octachron** [comment]
Thanks for the confirmation. Indeed rebase across repertory moves tend to be painful, sorry for the noise!

**Pull request merged**
